File: CSharp/SubscriptionApi/SubscriptionApi/Queries/HardwareQuery.cs
using System.Collections.Generic;
using System.Linq;
using SubscriptionApi.Authorization;

namespace SubscriptionApi.Queries
{
  public class HardwareQuery : AssetQuery
  {
    private readonly string _name;
    private readonly AssetQueryResolution _authorizationStructure;

    public HardwareQuery(string name, AssetQueryResolution authorizationStructure)
    {
      _name = name;
      _authorizationStructure = authorizationStructure;
    }

    public void ResolveInto(List<string> requestedAssetNames, QueryResolutionEvents resolutionEvents)
    {
      var assetsFromQuery = _authorizationStructure.RetrieveAssetsByHardwareName(_name);
      if (!assetsFromQuery.Any())
      {
        resolutionEvents.NoResolutionResultsFor(_name);
      }
      else
      {
        requestedAssetNames.AddRange(assetsFromQuery);
      }
    }
  }
}

File: Java/factories/_01_no_factories/src/main/java/Queries/HardwareQuery.java
package Queries;

import Authorization.AssetQueryResolution;

import java.util.List;

public class HardwareQuery implements AssetQuery {
    private final String _name;
    private final AssetQueryResolution _authorizationStructure;

    public HardwareQuery(String name, AssetQueryResolution authorizationStructure) {
        _name = name;
        _authorizationStructure = authorizationStructure;
    }

    public void ResolveInto(List<String> requestedAssetNames, QueryResolutionEvents resolutionEvents) {
        List<String> assetsFromQuery = _authorizationStructure.RetrieveAssetsByHardwareName(_name);
        if (assetsFromQuery.isEmpty()) {
            resolutionEvents.NoResolutionResultsFor(_name);
        } else {
            requestedAssetNames.addAll(assetsFromQuery);
        }
    }
}
